File: ruby/sinatra/spec/controllers/users/sessions_controller_spec.rb
RSpec.describe Users::SessionsController do
  let(:email) { "user@example.com" }
  let(:password) { "password" }

  describe "POST /users/sign_in" do
    let!(:user) { create(:user) }

    context "with proper params" do
      before do
        post "/users/sign_in", params: { user: { email:, password: } }
      end

      it "returns a JSON object" do
        expect(last_response.body).to be_a String
        expect(JSON.parse(last_response.body)).to have_key "id"
      end

      it "returns the instance of User" do
        data = JSON.parse(last_response.body)
        expect(data["id"]).to be user.id
      end

      it "returns a ok HTTP status" do
        expect(last_response.status).to be 200
      end

      it "returns a Authorization header" do
        expect(last_response["Authorization"]).to eq user.jwt
      end
    end

    context "without proper params" do
      before do
        post "/users/sign_in", params: { user: { email: nil, password: nil } }
      end

      it "returns a JSON object" do
        expect(last_response.body).to be_a String
        expect(JSON.parse(last_response.body)).to have_key "errors"
      end

      it "returns a list of error messages" do
        data = JSON.parse(last_response.body)
        expect(data["errors"]).to match({ "user" => "Invalid Email or Password" })
      end

      it "returns a unauthorized HTTP status" do
        expect(last_response.status).to be 401
      end
    end
  end

  describe "DELETE /users/sign_out" do
    context "when a User is authenticated" do
      let(:user) { create(:user) }

      before do
        delete "/users/sign_out", nil, { "HTTP_AUTHORIZATION" => user.jwt }
      end

      it "signs the User out" do
        get "/current_user", nil, { "HTTP_AUTHORIZATION" => user.jwt }
        expect(last_response.status).to be 401
      end

      it "returns a ok HTTP status" do
        expect(last_response.status).to be 200
      end
    end

    context "when a User is not authenticated" do
      it "returns a unauthorized HTTP status" do
        delete "/users/sign_out"

        expect(last_response.status).to be 401
      end
    end
  end
end

File: node/express/spec/controllers/users/sessions.controller.spec.ts
import supertest from "supertest"
import { beforeEach } from "vitest"
import app from "src/app.ts"
import prisma, { type User } from "src/models/user.model.ts"

describe("UsersSessionsControler", () => {
  let response: supertest.Response
  const request = supertest(app)
  const email = "user@example.com"
  const password = "password"
  let user: User

  const createUser = async () => await prisma.user.create({ data: { email, password } })

  beforeEach(async () => {
    user = await createUser()
  })

  describe("POST /users/sign_in", () => {
    describe("with proper params", () => {
      beforeEach(async () => {
        response = await request.post("/users/sign_in").send({ user: { email, password } })
      })

      it("returns a JSON object", () => {
        expect(response.type).toBe("application/json")
      })

      it("returns the instance of User", () => {
        expect(response.body).toMatchObject({ id: user.id })
      })

      it("returns a ok HTTP status", () => {
        expect(response.status).toBe(200)
      })

      it("returns a Authorization header", () => {
        expect(response.headers.authorization).toBeDefined()
        expect(response.headers.authorization).toBe(user.jwt)
      })
    })

    describe("without proper params", () => {
      beforeEach(async () => {
        response = await request.post("/users/sign_in").send({ user: { email: null, password: null } })
      })

      it("returns a JSON object", () => {
        expect(response.type).toBe("application/json")
      })

      it("returns a list of error messages", () => {
        expect(response.body).toStrictEqual({ errors: { user: "Invalid Email or Password" } })
      })

      it("returns a unauthoriazed HTTP status", () => {
        expect(response.status).toBe(401)
      })
    })
  })

  describe("DELETE  /users/sign_out", () => {
    describe("when a User is authenticated", () => {
      beforeEach(async () => {
        response = await request.delete("/users/sign_out").set("Authorization", user.jwt)
      })

      it("signs the User out", async () => {
        response = await request.get("/current_user").set("Authorization", user.jwt)
        expect(response.status).toBe(401)
      })

      it("returns a ok HTTP status", () => {
        expect(response.status).toBe(200)
      })
    })

    describe("when a User is not authenticated", () => {
      beforeEach(async () => {
        response = await request.delete("/users/sign_out")
      })

      it("returns a unauthorized HTTP status", () => {
        expect(response.status).toBe(401)
      })
    })
  })
})
